packages/regl-worldview/src/WorldviewContext.js

```diff
@@ -227,7 +232,16 @@ export class WorldviewContext {
     this.counters.render = Date.now() - start;
   }
 
-  _debouncedPaint = debounce(this.paint, 10);
+  tick = () => {
+    cancelAnimationFrame(this._frame);
+    this._frame = window.requestAnimationFrame(() => {
+      if (this._dirty) {
+        this.paint();
+        this._dirty = false;
+      }
+      this.tick();
```

Comment:
Was this recursive call intentional? A user of Worldview whose data never updates shouldn't have to pay the cost of re-rendering at 60fps.

Reply:
Yes it's intentional. They won't pay the cost, this will "tick" but not "paint" if the "dirty" flag has not been set since the last "paint".

 I added `console.log("tick")` and `console.log("paint")` to verify that if I call `onDirty` at 10fps, the loop "ticks" at 60fps, but paint only gets called at 10fps.

The cost of checking the boolean `dirty` flag at 60fps should be nominal

Reply:
I still think it's a waste of CPU that we should avoid

Reply:
Hmm, I'm surprised to hear that you feel that way, it's only taking 9 μs every 16ms, that's 9 millionths of a second of CPU only, and the overall changes improve performance for apps that use `onDirty`... What would you suggest instead?

Reply:
I think it just makes me sad when web apps do things on recurring timers or tight loops and it slows down my browser 😛 Keep in mind that some apps have several Worldviews on a single page.

Doesn't it still work if you just delete this line? If someone calls onDirty, that will lead to a paint call on the next available animation frame. If they don't, nothing else happens.

Might also be worth looking at debouncePromise: https://github.com/cruise-automation/webviz/blob/master/packages/webviz-core/src/util/debouncePromise.js

Reply:
I agree, but keep in mind that's why I'm using RAF, to ensure that consumers calling onDirty more frequently than the browser can handle does not cause any performance degradation.

Right now someone can call this every 12ms, and it can take 50ms each time to run, which will segfault the browser due to memory leaks / back pressure

This change will eliminate the possibility of the type of performance issue you're describing.

If we delete the recursion, this will then run 1x only. That means only the first map tile loading would cause a paint, and subsequent map tile loading would not cause a paint. We need this loop to keep checking if there's something new to paint, I'm pretty sure. Also like I'm saying since it's a RAF, it should adaptively be called less often in a CPU constrained browser, so I'm not personally concerned.

Re: debounced promise, this is a long lived loop so if you're looking for a more elegant abstraction I would recommend looking into RxJS with a RAF scheduler, since observables are like long lived promises that emit 0 to N times over time, unlike a promise which just emits once. RxJS has several built in throttle/debounce timers, but I'm contending we specifically don't want a debounce or throttle.

A throttle would bake in assumptions about how CPU constrained the browser is, or how much work is being done. It would still allow the method to be called more frequently than the browser can handle it, and crash the browser or degrade performance.

A debounce would mean if it gets called more frequently than the debounce interval, it just never paints.

The RAF loop avoids both problems, and ensures we never force feed the browser more work than it can handle, because we won't schedule another frame until the previous frame completes.

I can empathize with you here though, and could envision a middle-ground where we have `onDirty` sometimes paint synchronously, and sometimes schedule a RAF callback, depending on the delta between when it was last called vs when it's currently being called. It would a lot of complexity to the code, and would make little difference, in my opinion, but I'd be happy to make the change if you feel strongly.

Reply:
Disclaimer: I am not really an owner of this library anymore, so I think you should feel free to abandon this conversation and take someone else's opinion if you're blocked. I'm just speculating as a bystander. I find the conversation interesting but I don't want to slow you down too much :)

> If we delete the recursion, this will then run 1x only. That means only the first map tile loading would cause a paint, and subsequent map tile loading would not cause a paint. 

I don't think this is right in general. Suppose one map tile loads, and calls onDirty. If the second one loads before the animation frame, great, its onDirty will do nothing (or replace the current frame request, but that's effectively nothing) and the next animation frame will repaint both of them (effectively using the browser's animation timer as a way of batching the two updates). On the other hand, if the second one loads **after** the first animation frame, its onDirty call will just schedule another frame. That's fine too, since the paint calls won't actually happen faster than the browser can render them.

Since paint is a synchronous call, I'm ignoring the case of "the second one loads **during** the first animation frame". On the other hand, the [debouncePromise code](https://github.com/cruise-automation/webviz/blob/master/packages/webviz-core/src/util/debouncePromise.js) I wrote was meant to handle the case of an asynchronous bit of work, and that's why it has a second flag for `callPending`. With async work (paint), if a second call comes in while the first one is being serviced (painted), you need to give the second one another opportunity to run, but only after the first one has finished (since you don't want to be painting 2 things at once).

I admit `debouncePromise` was probably not the best name (I think I just forgot / didn't know the terminology of debounce vs throttle). I agree with your points about "classic" throttle and debounce both making assumptions about performance limits or leaving it open to perf degradations. My `debouncePromise` doesn't have a timer or animation frame, though, it exists only for the purpose of batching requests while ensuring there is always a trailing-edge call. I also realize I should've added more comments to it 😓

You might be able to tweak your current implementation pretty easily, since the Worldview paint() is synchronous. On the other hand you might be able to use debouncePromise easily too, like this:

```js
function rafPromise(): Promise<void> {
  return new Promise((resolve) => window.requestAnimationFrame(resolve));
}
debouncedPaint = debouncePromise(async () => {
  await rafPromise();
  paint();
});
```

Reply:
I like all the feedback I can get :) So no worries.

> If the second one loads after the first animation frame, its onDirty call will just schedule another frame

I think that's the part you may be missing here. `onDirty()` just toggles a boolean now, with my latest changes.  Then separately the RAF loop which is always running checks the `dirty` flag and conditionally paints. `onDirty()` isn't scheduling any work directly, just setting that flag for the RAF loop to read on it's next tick. Does that change your feedback?

Reply:
I see, yeah, I guess I'm arguing for going back to something like this:

```js
onDirty = () => {
  if (!this._frame) this._frame = requestAnimationFrame(this.paint);
}
componentDidUpdate() {
  if (!this._frame) this._frame = requestAnimationFrame(this.paint);
}
paint = () => {
  /*webgl rendering here*/
  this._frame = null;
}
```

Do you think this implementation would still be problematic for your use case? It seems like the best of both words, where the actual paints will never be delayed longer than necessary, and you can call onDirty as often as you like.

Reply:
I think that's solid feedback, I updated the PR and I can see in both the perf tab & through manual testing the perf is even better now. I of course had to modify your psuedo code to wrap this.paint in an arrow function so RAF doesn't set the context for `this` to the window object. Thanks for your feedback!